### DISCUSSION TITLE
ESP-NOW Low Latency Streaming of Audio

### DISCUSSION THREAD

**Discussion by gary7530** [pschatzmann/arduino-audio-tools#605] [Q&A] [answered]
I now have successful content, there are three ESP32, ESP1 is responsible for reading MP3 from SD card, and then send to ESP2 and ESP3 through ESP NOW.

The problem is that the music played by ESP2 and ESP3 is not in sync, how can I solve this problem.

The example I use is "communication-espnow-send" with "communication-espnow-receive"

**Comment by pschatzmann** [comment, marked as answer]
Actually, you can't: they will always be out of sync because they might not start at the same time or the processors have slightly different clock speeds.  So the challenge is to minimise this difference so that it ins not disturbing too much.

The easiest solution is just to shut the door between the rooms, so that it does not disturb!

The best  that you can try to do is to fill some buffer on the receiving system and add some logic to make sure that they both start to play at the same time and re-sychronize at the start of the next title.

How far are they currently apart ?

**Reply by gary7530** [reply]
"The best thing you can try to do is fill some buffers on the receiving system and add some logic to make sure they all start playing at the same time and mentally adjust again at the start of the next title."  has this paragraph Is there any relevant example of the code of the language?

The distance is about 10 meters.

**Comment by pschatzmann** [comment]
Oh, sorry I mean distance in time: Is it a couple of seconds or half a second...

**Reply by gary7530** [reply]
About 0.5~1 second

**Comment by gary7530** [comment]
I'm using "StreamCopy" to play music on the receiving end, and I want to change it to "AudioPlayer", so how do I change it?

This is my code now:
```
AnalogAudioStream out;
EncodedAudioStream decoder(&out, new MP3DecoderHelix());
ESPNowStream now;
StreamCopy copier(decoder, now);
```
----------------------------------------------------------------------------------
I changed the code:
```
AnalogAudioStream out;
MP3DecoderHelix decoder;
ESPNowStream now;
AudioPlayer player(now, out, decoder);
```

An error will occur:
```
C:\Users\user\GoogleDrive\Code\Arduino\MP3\arduino-audio-tools\examples\examples-communication\esp-now\communication-espnow-receive\communication-espnow-receive.ino:27:37: error: no matching function for call to 'audio_tools::AudioPlayer::AudioPlayer(audio_tools::ESPNowStream&, audio_tools::AnalogAudioStream&, audio_tools::MP3DecoderHelix&)'
 AudioPlayer player(now, out, decoder);
                                     ^
In file included from c:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools.h:78,
                 from C:\Users\user\GoogleDrive\Code\Arduino\MP3\arduino-audio-tools\examples\examples-communication\esp-now\communication-espnow-receive\communication-espnow-receive.ino:11:
c:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioPlayer.h:438:9: note: candidate: 'audio_tools::AudioPlayer::AudioPlayer()'
         AudioPlayer() {
         ^~~~~~~~~~~
c:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioPlayer.h:438:9: note:   candidate expects 0 arguments, 3 provided
In file included from c:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools.h:78,
                 from C:\Users\user\GoogleDrive\Code\Arduino\MP3\arduino-audio-tools\examples\examples-communication\esp-now\communication-espnow-receive\communication-espnow-receive.ino:11:
c:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioPlayer.h:99:9: note: candidate: 'audio_tools::AudioPlayer::AudioPlayer(audio_tools::AudioSource&, audio_tools::AudioStream&, audio_tools::AudioDecoder&)'
         AudioPlayer(AudioSource& source, AudioStream& output, AudioDecoder& decoder) {
         ^~~~~~~~~~~
c:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioPlayer.h:99:9: note:   no known conversion for argument 1 from 'audio_tools::ESPNowStream' to 'audio_tools::AudioSource&'
c:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioPlayer.h:77:9: note: candidate: 'audio_tools::AudioPlayer::AudioPlayer(audio_tools::AudioSource&, Print&, audio_tools::AudioDecoder&, audio_tools::AudioBaseInfoDependent*)'
         AudioPlayer(AudioSource& source, Print& output, AudioDecoder& decoder, AudioBaseInfoDependent* notify = nullptr) {
         ^~~~~~~~~~~
c:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioPlayer.h:77:9: note:   no known conversion for argument 1 from 'audio_tools::ESPNowStream' to 'audio_tools::AudioSource&'
c:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioPlayer.h:51:9: note: candidate: 'audio_tools::AudioPlayer::AudioPlayer(audio_tools::AudioSource&, audio_tools::AudioPrint&, audio_tools::AudioDecoder&)'
         AudioPlayer(AudioSource& source, AudioPrint& output, AudioDecoder& decoder) {
         ^~~~~~~~~~~
c:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioPlayer.h:51:9: note:   no known conversion for argument 1 from 'audio_tools::ESPNowStream' to 'audio_tools::AudioSource&'

exit status 1

Compilation error: no matching function for call to 'audio_tools::AudioPlayer::AudioPlayer(audio_tools::ESPNowStream&, audio_tools::AnalogAudioStream&, audio_tools::MP3DecoderHelix&)'
```

**Reply by isadorabit** [reply]
Hi!
I'm trying to do the exact same thing. 
did you manage to solve the sync problem between ESP2 and ESP3??
if yes, how?

**Reply by gary7530** [reply]
Hello, I did not solve the problem of esp now delay

**Reply by gcopeland** [reply]
Just FYI, if you want this to work, you MUST use espnow's broadcast mechanism (all peer address octets are 0xff). Otherwise, the next node will always be time shifted by the time it takes to send to all previous nodes. 

I didn't see the ESPNowStream code, but based on the reported delays, I'm assuming you're using sequential unicast to multiple receivers. You don't want to do this with realtime data and multiple receivers. And even then, because of retries and error correction, even with one receiver this can become real iffy. This is why you usually use UDP on IP networks for this type of thing. This is further desynced by any retransmission delays for each node. The only way to be reasonably close in time here (because of distances between tx and each rx), is to broadcast such that the data is sent only once and all receivers receive it at the same time (delta for time-distance).

Given fairly short distances, any time deltas created by distance should be relatively minor and not enough to notice to the ear.

Also note, as with UDP, broadcast mode disables error correction and retransmissions. Which means it's possible from time to time a rx node will not receive the audio for the given time slice. When this happens, you will hear it.

**Comment by pschatzmann** [comment]
I am not sure if this makes any sense, but I suggest that you look at the class documentation to figure out what you need to provide to the constructor.
The first parameter is an AudioSource and there is no implementation for this which would be based on esp-now. So you would need to write this yourself! 

It should not be too difficult to implement this...
